### PR TITLE
Add support for CPR (and variants) for flow_ebos

### PR DESCRIPTION
**Pull request by blattms**
The approach is slightly different to the one in flow_ebos:

1. We scale the linear systems a bit differently. flow_ebos uses a simple decoupling approach proposed by  Scheichl and Masson while flow_legacy uses the DRS left preconditioning of Gries (supposedly to make the pressure system more adequate for AMG). DRS could be added as option here.
2. flow_legacy also scales the pressure components in addition for all solver approaches.
3. For flow_ebos one can also aggregate the pressure system (in addition to extracting it) and use that as the coarse system.

CPR converges for flow_ebos (broken for flow_legacy) but as for flow_legacy the simulation is slower than for plain ILU.

As all diligent reviewers will note, this PR also switches CPRPreconditioner.hpp to OverlappingParallelILU0 and deletes some unneeded code (AdditionalObjectDeleter.hpp). This is done in 6d21214fa and could lso be a separate PR.

Before merging this I have to test in parallel again as I might broken something in the meantime.

### CODE REVIEW

**Comment by blattms** [comment]
jenkins build this please

File: opm/autodiff/CPRPreconditioner.hpp [hunk 1]
```diff
@@ -67,209 +72,138 @@ template<class M, class X, class Y, class P>
...
         ::EllipticPreconditionerPointer EllipticPreconditionerPointer;
     return EllipticPreconditionerPointer(new ParallelPreconditioner(Ae, comm, relax));
 }
 
-#endif // #if HAVE_MPI
+template < class C, class Op, class P, class S, std::size_t index >
+inline void
+createAMGPreconditionerPointer(Op& opA, const double relax, const P& comm,
+                               std::unique_ptr< BlackoilAmg<Op,S,C,P,index> >& amgPtr,
+                               const CPRParameter& params = CPRParameter())
```
Comment (atgeirr) [inline opm/autodiff/CPRPreconditioner.hpp]: I get lots of compile errors trying to compile this PR. The first one is here: invalid use of incomplete type 'Opm::CPRParameter'. That error is repeated multiple times. I also get: no matching function for call to 'createAMGPreconditionerPointer', and similar errors. Forgotten #include?

File: opm/autodiff/BlackoilAmg.hpp [hunk 2]
```diff
@@ -828,20 +841,39 @@ class OneComponentAggregationLevelTransferPolicy
     bool cpr_pressure_aggregation_;
 };
 
+/**
+ * \brief An algebraic twolevel or multigrid approach for solving blackoil (supports CPR with and without AMG)
+ *
+ * This preconditioner first decouples the component used for coarsening using a simple scaling
+ * approach (e.g. Scheichl, Masson 2013,\see scaleMatrixQuasiImpes). Then it constructs the first
+ * coarse level system, either by simply extracting the coupling between the components at COMPONENT_INDEX
+ * in the matrix blocks or by extracting them and applying aggregation to the directly. This coarse level
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: s/to the/g or the likes.

File: opm/autodiff/BlackoilAmg.hpp [hunk 3]
```diff
@@ -828,20 +841,39 @@ class OneComponentAggregationLevelTransferPolicy
...
+ * can be solved either by AMG or by ILU. The preconditioner is configured using CPRParameter.
+ * \tparam O The type of the operator (encapsulating a BCRSMatrix).
+ * \tparam S The type of the smoother.
+ * \tparam C The type of coarsening criterion to use.
+ * \tparam P The type of the class describing the parallelelization.
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: l's and e's run amok :)

File: opm/autodiff/BlackoilAmg.hpp [hunk 4]
```diff
@@ -828,20 +841,39 @@ class OneComponentAggregationLevelTransferPolicy
...
 class BlackoilAmg
     : public Dune::Preconditioner<typename O::domain_type, typename O::range_type>
 {
 public:
+    /** \brief The type of the operator (encapsulating a BCRSMatrix). */
     using Operator = O;
-    using Matrix = typename Operator::matrix_type;
+    /** \brief The type of coarsening criterion to use. */
     using Criterion = C;
+    /** \brief The type of the class describing the parallelelization. */
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: parallelization

File: opm/autodiff/BlackoilAmg.hpp [hunk 5]
```diff
@@ -0,0 +1,727 @@
...
+            {
+                amg_.pre(x,b);
+                first_=false;
+                x_=x;
+            }
+            amg_.apply(x,b);
+            */
+            using Chooser = Dune::ScalarProductChooser<X,Communication,AMGType::category>;
+            auto sp = Chooser::construct(comm_);
+            Dune::BiCGSTABSolver<X> solver(const_cast<typename AMGType::Operator&>(op_), *sp, amg_, 1e-2, 25, 0);
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: hardcoded params, possibly forgotten?

File: opm/autodiff/BlackoilAmg.hpp [hunk 6]
```diff
@@ -0,0 +1,727 @@
...
+        AMGInverseOperator(const AMGInverseOperator& other)
+            : x_(other.x_), amg_(other.amg_), first_(other.first_)
+        {
+        }
+    private:
+        X x_;
+        AMGType amg_;
+        const typename AMGType::Operator& op_;
+        const Communication& comm_;
+        bool first_;
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: unused (due to commented code).

File: opm/autodiff/BlackoilAmg.hpp [hunk 7]
```diff
@@ -0,0 +1,727 @@
...
+                     const S& overlap,
+                     std::size_t& overlapCount)
+{
+    // count the overlap vertices.
+    auto end = graph.end();
+    overlapCount = 0;
+
+    const auto& lookup=pinfo.globalLookup();
+
+    for(auto vertex=graph.begin(); vertex != end; ++vertex) {
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: could use range for

File: opm/autodiff/BlackoilAmg.hpp [hunk 8]
```diff
@@ -0,0 +1,727 @@
...
+        const auto* pair = lookup.pair(*vertex);
+
+        if(pair!=0 && overlap.contains(pair->local().attribute()))
+            ++overlapCount;
+    }
+    // Allocate space
+    using Vertex =  typename G::VertexDescriptor;
+    using OverlapVertex = Dune::Amg::OverlapVertex<Vertex>;
+
+    auto* overlapVertices = new OverlapVertex[overlapCount=0 ? 1 : overlapCount];
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: == intended?

File: opm/autodiff/BlackoilAmg.hpp [hunk 9]
```diff
@@ -0,0 +1,727 @@
...
+    if(overlapCount==0)
+        return overlapVertices;
+
+    // Initialize them
+    overlapCount=0;
+    for(auto vertex=graph.begin(); vertex != end; ++vertex) {
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: could use range for

File: opm/autodiff/BlackoilAmg.hpp [hunk 10]
```diff
@@ -0,0 +1,727 @@
...
+
+    // Reset the visited flags of all vertices.
+    // As the isolated nodes will be skipped we simply mark them as visited
+#ifndef NDEBUG
+    const auto UNAGGREGATED = Dune::Amg::AggregatesMap<typename G::VertexDescriptor>::UNAGGREGATED;
+#endif
+    const auto ISOLATED = Dune::Amg::AggregatesMap<typename G::VertexDescriptor>::ISOLATED;
+    auto vend = fineGraph.end();
+
+    for(auto vertex = fineGraph.begin(); vertex != vend; ++vertex) {
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: could use range for

File: opm/autodiff/BlackoilAmg.hpp [hunk 11]
```diff
@@ -0,0 +1,727 @@
...
+{
+    // Reset the visited flags of all vertices.
+    // As the isolated nodes will be skipped we simply mark them as visited
+#ifndef NDEBUG
+    const auto UNAGGREGATED = Dune::Amg::AggregatesMap<typename G::VertexDescriptor>::UNAGGREGATED;
+#endif
+    const auto ISOLATED = Dune::Amg::AggregatesMap<typename G::VertexDescriptor>::ISOLATED;
+    using Vertex = typename G::VertexIterator;
+    Vertex vend = fineGraph.end();
+    for(Vertex vertex = fineGraph.begin(); vertex != vend; ++vertex) {
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: could use range for

File: opm/autodiff/BlackoilAmg.hpp [hunk 12]
```diff
@@ -0,0 +1,727 @@
...
+                                                 *communication_, OverlapFlags());
+
+        aggregatesMap_.reset(new AggregatesMap(std::get<1>(graphs)->maxVertex()+1));
+
+        int noAggregates, isoAggregates, oneAggregates, skippedAggregates;
+        using std::get;
+        std::tie(noAggregates, isoAggregates, oneAggregates, skippedAggregates) =
+            aggregatesMap_->buildAggregates(fineOperator.getmat(), *(get<1>(graphs)),
+                                            criterion_, true);
+        std::cout<<"no aggregates="<<noAggregates<<" iso="<<isoAggregates<<" one="<<oneAggregates<<" skipped="<<skippedAggregates<<std::endl;
```
Comment (akva2) [inline opm/autodiff/BlackoilAmg.hpp]: cout logging wanted?

File: opm/autodiff/CPRPreconditioner.hpp [hunk 13]
```diff
@@ -271,6 +271,29 @@ createEllipticPreconditionerPointer(const M& Ae, double relax,
...
+    // TODO: revise choice of parameters
+    int coarsenTarget=1200;
+    using Criterion = C;
+    Criterion criterion(15, coarsenTarget);
+    criterion.setDebugLevel( 0 ); // no debug information, 1 for printing hierarchy information
+    criterion.setDefaultValuesIsotropic(2);
+    criterion.setNoPostSmoothSteps( 1 );
+    criterion.setNoPreSmoothSteps( 1 );
+
+    // for DUNE 2.2 we also need to pass the smoother args
```
Comment (akva2) [inline opm/autodiff/CPRPreconditioner.hpp]: maybe not necessary any longer ?

File: opm/autodiff/ISTLSolver.hpp [hunk 14]
```diff
@@ -420,12 +419,34 @@ namespace Opm
                 }
 
                 const double relax = 1.0;
+                if ( ! parameters_.amg_blackoil_system_ )
```
Comment (akva2) [inline opm/autodiff/ISTLSolver.hpp]: personal preference: switch else case.

File: opm/autodiff/CPRPreconditioner.hpp (hunk 1 above)
Comment (blattms) [reply opm/autodiff/CPRPreconditioner.hpp]: Weired that my compilers (g++ 6.3.18 and 4.9.2) did not complain. But yours is right constructing CPRParameter needs the definition which only available at the end of the file.

Comment (blattms) [reply opm/autodiff/CPRPreconditioner.hpp]: Weired that my compilers (g++ 6.3.18 and 4.9.2) did not complain. But yours is right constructing CPRParameter needs the definition which only available at the end of the file.

File: opm/autodiff/BlackoilAmg.hpp (hunk 5 above)
Comment (blattms) [reply opm/autodiff/BlackoilAmg.hpp]: I guess this was fixed in a later commit (e24d338d)

File: opm/autodiff/CPRPreconditioner.hpp (hunk 13 above)
Comment (blattms) [reply opm/autodiff/CPRPreconditioner.hpp]: already fixed in later commit.

Comment (blattms) [reply opm/autodiff/CPRPreconditioner.hpp]: done with next push

File: opm/autodiff/BlackoilAmg.hpp (hunk 12 above)
Comment (blattms) [reply opm/autodiff/BlackoilAmg.hpp]: comment out later and removed with next push

File: opm/autodiff/BlackoilAmg.hpp (hunk 8 above)
Comment (blattms) [reply opm/autodiff/BlackoilAmg.hpp]: Nope. Thanks for the catch.

File: opm/autodiff/CPRPreconditioner.hpp (hunk 1 above)
Comment (blattms) [reply opm/autodiff/CPRPreconditioner.hpp]: Fixed.

Comment (blattms) [reply opm/autodiff/CPRPreconditioner.hpp]: At least that is what I guess. Please try again.

**Comment by blattms** [comment]
Thanks for the review. Tried my best to incorporate your suggestions.

File: opm/autodiff/CPRPreconditioner.hpp [hunk 15]
```diff
@@ -67,234 +115,124 @@ template<class M, class X, class Y, class P>
...
 
-    // The coarsening criterion used in the AMG
-    typedef Dune::Amg::CoarsenCriterion<CritBase> Criterion;
+    amgPtr.reset( new AMG( params, opA, criterion, smootherArgs, comm ) );
+}
 
+template < class C, class Op, class P, class AMG >
+inline void
+createAMGPreconditionerPointer(Op& opA, const double relax, const P& comm, std::unique_ptr< AMG >& amgPtr,
+                               const CPRParameter& params = CPRParameter())
```
Comment (atgeirr) [inline opm/autodiff/CPRPreconditioner.hpp]: This parameter is unused. Did you intend to use it to control what is now hardcoded below?

File: opm/autodiff/ParallelOverlappingILU0.hpp [hunk 16]
```diff
@@ -393,6 +420,22 @@ class ParallelOverlappingILU0
...
     void init( const Matrix& A, const int iluIteration )
     {
+        // (For older DUNE versions the communicator might be
+        // invalid if redistribution in AMG happened on the coarset level.
+        // Therefore we check for nonzero size
+        if ( comm_ && comm_->communicator().size()<=0 )
+        {
+            if ( A.N() > 0 )
+            {
+                OPM_THROW(std::logic_error, "Expected a matrix with zero rows for an invalid communicator.");
```
Comment (atgeirr) [inline opm/autodiff/ParallelOverlappingILU0.hpp]: This requires including opm/common/ErrorMacros.hpp.

There seems to be a commented-out OPM_THROW in apply() around line 360 that could have been disabled for the same reason? Right now it uses std::abort() instead which seems a little harsh.

File: tests/test_blackoil_amg.cpp [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+    typedef typename Dune::Amg::SmootherTraits<ParSmoother>::Arguments SmootherArgs;
+
+    Dune::OverlappingSchwarzScalarProduct<Vector,Communication> sp(comm);
+
+    Dune::InverseOperatorResult r;
+    SmootherArgs smootherArgs;
+    Criterion criterion;
+
+    smootherArgs.iterations = 1;
+    Opm::CprParameter param;
```
Comment (atgeirr) [inline tests/test_blackoil_amg.cpp]: Should be CPRParameter. Weird that this compiled?!?

File: tests/test_blackoil_amg.cpp [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+}
+
+bool init_unit_test_func()
+{
+    return true;
+}
+
+int main(int argc, char** argv)
+{
+    const auto& helper = Dune::MPIHelper::instance(argc, argv);
```
Comment (atgeirr) [inline tests/test_blackoil_amg.cpp]: Unused variable.

**Comment by blattms** [comment]
Nice catch

**Comment by blattms** [comment]
All of @atgeirr 's comments have been adressed. Performance is still lousy for
Norne.

**Comment by blattms** [comment]
Because of the performance problems I also tried the scaling using the DRS preconditioner. I guess that is what was implemented for flow_legacy. It turns out that for the matrix that our preconditioner gets (no well contributions) the resulting matrix after the scaling is the same as with the scaling of Scheichl & Masson.

I will try whether this changes if the well contributions are added, i.e. PR #1183.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
Noticed a single warning when compiling:
```
/Users/atgeirr/opm/opm-simulators/tests/test_blackoil_amg.cpp:21:9: warning: 'BOOST_TEST_DYN_LINK' macro redefined [-Wmacro-redefined]
#define BOOST_TEST_DYN_LINK
        ^
<command line>:1:9: note: previous definition is here
#define BOOST_TEST_DYN_LINK 1
        ^
```
We now rely on the build system to set this, see PR #1406 that adapted the other tests.

**Comment by atgeirr** [comment]
All is green, I will merge now. The feature should be considered experimental, we will have to do more work on it to make it worth bragging about (but we should get there...)

The minor BOOST_TEST_DYN_LINK warning I'll fix myself in a separate PR.

**Pull request merged**